src/ds/rb_map.h

```diff
@@ -284,6 +303,353 @@ namespace rb
+          root.rootValue(),
+          fixedRight.first);
+        if (!fixedRight.second)
+        {
+          // nothing left to fix
+          return std::make_pair(root, false);
+        }
+        // in fixing that we may have moved the double black to a child of this
+        // root, so fix that now
+        sibling = root.left();
+      }
+
+      auto doubleBlack = false;
+      if (sibling.left().rootColor() == B && sibling.right().rootColor() == B)
+      {
+        // current node is being made black, siblings children are both black so
+        // we can safely convert the sibling to red and propagate the double
+        // black
+        sibling = sibling.paint(R);
+        // we might still have to propagate the double black up
+        doubleBlack = root.rootColor() == B;
+        root = Map(B, sibling, root.rootKey(), root.rootValue(), root.right());
+      }
+      else
+      {
+        if (sibling.left().rootColor() == B)
+        {
+          // root, sibling and sibling's left are all black
+          // rotate the right with the sibling as the root
+          auto siblingRight = sibling.right().paint(B);
+          sibling = Map(
+            R,
+            sibling.left(),
+            sibling.rootKey(),
+            sibling.rootValue(),
+            siblingRight);
+          sibling = sibling.rotateLeft();
+          root = Map(
+            root.rootColor(),
+            sibling,
+            root.rootKey(),
+            root.rootValue(),
+            root.right());
+        }
+
+        auto recoloredSibling = Map(
+          root.rootColor(),
+          sibling.left().paint(B),
+          sibling.rootKey(),
+          sibling.rootValue(),
+          sibling.right());
+        root = Map(
+          B, recoloredSibling, root.rootKey(), root.rootValue(), root.right());
+        root = root.rotateRight();
+        doubleBlack = false;
+      }
+      return std::make_pair(root, doubleBlack);
+    }
+
+    // Remove the node with the given key.
+    // returns a new map along with a bool indicating whether we need to handle
+    // a double black.
+    std::pair<Map, bool> _remove(const K& key) const
+    {
+      if (empty())
+      {
+        // key not present in the tree, can't remove it so just return an empty
+        // map.
+        return std::make_pair(Map(), false);
+      }
+
+      auto& rootk = rootKey();
+
+      if (key < rootk)
+      {
+        // remove key from the left subtree
+        auto left_without = left()._remove(key);
+        // copy the left into a new map to return
+        auto newMap =
+          Map(rootColor(), left_without.first, rootKey(), rootValue(), right());
+        if (left_without.second)
+        {
+          // there is a double black node in the left subtree so fix it up
+          return newMap.fixDoubleBlackLeft();
+        }
+        // no double blacks are present
+        return std::make_pair(newMap, false);
+      }
+      else if (rootk < key)
```

Comment:
```suggestion
      else if (key > rootk)
```

Just to keep the order of the operands consistent throughout the `if else`.

Reply:
That gives me an error about `invalid operands to binary expression`.

`&key > &rootk` works though...

Reply:
That breaks things so I'll leave it as it was.